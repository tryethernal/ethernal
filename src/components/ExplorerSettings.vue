<template>
    <v-card outlined>
        <v-card-text>
            <v-alert v-if="successMessage" dense text type="success">{{ successMessage }}</v-alert>
            <v-alert v-if="errorMessage" dense text type="error">{{ errorMessage }}</v-alert>
            <v-form @submit.prevent="updateExplorerSettings()" v-model="valid">
                <v-row>
                    <v-col>
                        <v-select
                            outlined
                            dense
                            label="Associated Workspace"
                            v-model="currentWorkspace"
                            item-text="name"
                            :items="workspaces"
                            return-object>
                            <template v-slot:item="{ item }">
                                {{ item.name }}<small class="ml-2">({{ shortRpcUrl(item.rpcServer) }} | {{  item.networkId }})</small>
                            </template>
                            <template v-slot:selection="{ item }">
                                {{ item.name }}<small class="ml-2">({{ shortRpcUrl(item.rpcServer) }} | {{  item.networkId }})</small>
                            </template>
                        </v-select>
                        <v-text-field
                            dense
                            outlined
                            v-model="explorer.name"
                            label="Name"></v-text-field>
                        <v-text-field
                            class="mb-2"
                            dense
                            outlined
                            v-model="explorer.slug"
                            :suffix="`.${mainDomain}`"
                            hint="Your explorer will always be reachable at this address"
                            persistent-hint
                            label="Ethernal Domain"></v-text-field>
                        <v-text-field
                            class="mb-2"
                            dense
                            outlined
                            :disabled="!capabilities.nativeToken"
                            :hint="capabilities.nativeToken ? '' : 'Upgrade your plan to customize your native token symbol.'"
                            v-model="explorer.token"
                            persistent-hint
                            label="Native Token Symbol"></v-text-field>
                        <v-text-field
                            dense
                            outlined
                            type="number"
                            :disabled="!capabilities.totalSupply"
                            :hint="capabilities.totalSupply ? `In ether: ${formatTotalSupply()}` : 'Upgrade your plan to display a total supply.'"
                            persistent-hint
                            hide-details="auto"
                            v-model="explorer.totalSupply"
                            label="Total Supply (in wei)"></v-text-field>
                    </v-col>
                </v-row>
                <v-card-actions>
                    <v-spacer></v-spacer>
                    <v-btn :loading="loading" color="primary" :disabled="!valid" type="submit">Update</v-btn>
                </v-card-actions>
            </v-form>
        </v-card-text>
    </v-card>
</template>

<script>
import { mapGetters } from 'vuex';
import { formatNumber, shortRpcUrl } from '../lib/utils';

export default {
    name: 'Explorer',
    props: ['explorer', 'workspaces'],
    data: () => ({
        successMessage: null,
        errorMessage: null,
        currentWorkspace: null,
        valid: false,
        loading: false,
        capabilities: {},
    }),
    mounted() {
        if (this.explorer.stripeSubscription)
            this.capabilities = this.explorer.stripeSubscription.stripePlan.capabilities;
    },
    methods: {
        shortRpcUrl,
        formatTotalSupply() {
            if (!this.explorer.totalSupply) return 'N/A';
            return formatNumber(this.explorer.totalSupply)
        },
        updateExplorerSettings() {
            this.loading = true;
            this.successMessage = null;
            this.errorMessage = null;
            const settings = {
                workspace: this.currentWorkspace.name,
                name: this.explorer.name,
                slug: this.explorer.slug
            };

            if (this.capabilities.nativeToken)
                settings['token'] = this.explorer.token;

            if (this.capabilities.totalSupply)
                settings['totalSupply'] = this.explorer.totalSupply;

            this.server.updateExplorerSettings(this.explorer.id, settings)
                .then(() => {
                    this.successMessage = 'Settings updated.';
                    this.$emit('updated');
                })
                .catch(error => {
                    this.errorMessage = error.response && error.response.data || 'Error while updating settings. Please retry.';
                })
                .finally(() => this.loading = false);
        }
    },
    computed: {
        ...mapGetters([
            'mainDomain'
        ]),
    },
    watch: {
        workspaces: {
            immediate: true,
            handler() {
                this.currentWorkspace = this.workspaces.find(w => w.id == this.explorer.workspaceId);
            }
        }
    }
}
</script>
