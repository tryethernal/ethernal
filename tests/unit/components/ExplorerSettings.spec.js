import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

import ExplorerSettings from '@/components/ExplorerSettings.vue';

let helper;
beforeEach(() => {
    jest.clearAllMocks()
    helper = new MockHelper();
});

const explorer = {
    workspaceId: 1,
    name: 'Explorer',
    slug: 'explorer',
    token: 'ETL',
    stripeSubscription: {
        stripePlan: {
            capabilities: { nativeToken: true, totalSupply: true }
        }
    }
};
const workspaces = [{ id: 1, name: 'Workspace 1', rpcServer: 'rpc', networkId: 1 }, { id: 2, name: 'Workspace 2', rpcServer: 'rpc', networkId: 1 }];

describe('ExplorerSettings.vue', () => {
    it('Should display settings with all options', async () => {
        const wrapper = helper.mountFn(ExplorerSettings, {
            propsData: { explorer, workspaces }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display settings with deactivated options', async () => {
        const wrapper = helper.mountFn(ExplorerSettings, {
            propsData: {
                explorer: { ...explorer, stripeSubscription: { stripePlan: { capabilities: { nativeToken: false, totalSupply: false }}}},
                workspaces
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });
});
