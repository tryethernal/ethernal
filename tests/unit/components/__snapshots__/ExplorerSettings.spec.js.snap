// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExplorerSettings.vue Should display settings with all options 1`] = `
<div class="v-card v-sheet v-sheet--outlined theme--light">
  <div class="v-card__text">
    <!---->
    <!---->
    <form novalidate="novalidate" class="v-form">
      <div class="row">
        <div class="col">
          <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined v-select">
            <div class="v-input__control">
              <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-6" class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-select__slot"><label for="input-6" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Associated Workspace</label>
                  <div class="v-select__selections">
                    Workspace 1<small class="ml-2">(rpc | 1)</small><input id="input-6" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                  </div><input type="hidden" value="[object Object]">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
          <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
            <div class="v-input__control">
              <div class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-text-field__slot"><label for="input-11" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Name</label><input id="input-11" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
          <div class="v-input mb-2 v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
            <div class="v-input__control">
              <div class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-text-field__slot"><label for="input-14" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Ethernal Domain</label><input id="input-14" type="text">
                  <div class="v-text-field__suffix">.tryethernal.com</div>
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                    <div class="v-messages__message">Your explorer will always be reachable at this address</div>
                  </transition-group-stub>
                </div>
              </div>
            </div>
          </div>
          <div class="v-input mb-2 v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
            <div class="v-input__control">
              <div class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-text-field__slot"><label for="input-17" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Native Token Symbol</label><input id="input-17" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
          <div class="v-input v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined">
            <div class="v-input__control">
              <div class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-text-field__slot"><label for="input-20" class="v-label theme--light" style="left: 0px; position: absolute;">Total Supply (in wei)</label><input id="input-20" type="number"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                    <div class="v-messages__message">In ether: N/A</div>
                  </transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-card__actions">
        <div class="spacer"></div> <button type="submit" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">Update</span></button>
      </div>
    </form>
  </div>
</div>
`;

exports[`ExplorerSettings.vue Should display settings with deactivated options 1`] = `
<div class="v-card v-sheet v-sheet--outlined theme--light">
  <div class="v-card__text">
    <!---->
    <!---->
    <form novalidate="novalidate" class="v-form">
      <div class="row">
        <div class="col">
          <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined v-select">
            <div class="v-input__control">
              <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-30" class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-select__slot"><label for="input-30" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Associated Workspace</label>
                  <div class="v-select__selections">
                    Workspace 1<small class="ml-2">(rpc | 1)</small><input id="input-30" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                  <div class="v-input__append-inner">
                    <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                  </div><input type="hidden" value="[object Object]">
                </div>
                <div class="v-menu">
                  <!---->
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
          <div class="v-input v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined">
            <div class="v-input__control">
              <div class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-text-field__slot"><label for="input-35" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Name</label><input id="input-35" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                </div>
              </div>
            </div>
          </div>
          <div class="v-input mb-2 v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined">
            <div class="v-input__control">
              <div class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-text-field__slot"><label for="input-38" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Ethernal Domain</label><input id="input-38" type="text">
                  <div class="v-text-field__suffix">.tryethernal.com</div>
                </div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                    <div class="v-messages__message">Your explorer will always be reachable at this address</div>
                  </transition-group-stub>
                </div>
              </div>
            </div>
          </div>
          <div class="v-input mb-2 v-input--is-label-active v-input--is-dirty v-input--is-disabled v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined">
            <div class="v-input__control">
              <div class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-text-field__slot"><label for="input-41" class="v-label v-label--active v-label--is-disabled theme--light" style="left: 0px; position: absolute;">Native Token Symbol</label><input disabled="disabled" id="input-41" type="text"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                    <div class="v-messages__message">Upgrade your plan to customize your native token symbol.</div>
                  </transition-group-stub>
                </div>
              </div>
            </div>
          </div>
          <div class="v-input v-input--is-disabled v-input--dense theme--light v-text-field v-text-field--is-booted v-text-field--enclosed v-text-field--outlined">
            <div class="v-input__control">
              <div class="v-input__slot">
                <fieldset aria-hidden="true">
                  <legend style="width: 0px;"><span class="notranslate">​</span></legend>
                </fieldset>
                <div class="v-text-field__slot"><label for="input-44" class="v-label v-label--is-disabled theme--light" style="left: 0px; position: absolute;">Total Supply (in wei)</label><input disabled="disabled" id="input-44" type="number"></div>
              </div>
              <div class="v-text-field__details">
                <div class="v-messages theme--light">
                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper">
                    <div class="v-messages__message">Upgrade your plan to display a total supply.</div>
                  </transition-group-stub>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-card__actions">
        <div class="spacer"></div> <button type="submit" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary"><span class="v-btn__content">Update</span></button>
      </div>
    </form>
  </div>
</div>
`;
